### PR TITLE
Port user|room|event|bridge stores to Typescript

### PR DESCRIPTION
**Pull request by Half-Shot**
This is a rather gargantuan PR, but it's mostly just typescriptification. The tests were modified to remove dependence on Bluebird, so this change is **breaking**, as we use pure promises here now.

### CODE REVIEW

File: src/components/event-bridge-store.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        return this.delete({
+            id: event.getId(),
+        });
+    }
+
+    /**
+     * Remove entries based on the matrix IDs.
+     * @param {string} roomId The ID of the room.
+     * @param {string} eventId The ID of the event.
+     * @return {Promise}
```
Comment (jaller94) [inline src/components/event-bridge-store.ts:84]:
```suggestion
     * @param roomId The ID of the room.
     * @param eventId The ID of the event.
```

File: src/components/event-bridge-store.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            "matrix.roomId": roomId,
+            "matrix.eventId": eventId,
+        });
+    }
+
+    /**
+     * Remove entries based on the matrix IDs.
+     * @param {string} roomId The ID of the room.
+     * @param {string} eventId The ID of the event.
+     * @return {Promise}
```
Comment (jaller94) [inline src/components/event-bridge-store.ts:97]:
```suggestion
     * @param roomId The ID of the room.
     * @param eventId The ID of the event.
```

File: src/components/event-bridge-store.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import Datastore from "nedb";
+import { BridgeStore } from "./bridge-store";
+import { StoredEvent, StoredEventDoc } from "../models/events/event";
+
+/**
+ * Construct a store suitable for event mapping information. Data is stored
+ * as {@link StoredEvent}s.
+ * @constructor
+ * @param {Datastore} db The connected NEDB database instance
```
Comment (jaller94) [inline src/components/event-bridge-store.ts:24]:
```suggestion
 * @param db The connected NEDB database instance
```

**Review by jaller94 — APPROVED** [review]
Looks good to me.

File: src/components/room-bridge-store.ts [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+     * <p>If a unique 'id' is not given, the store will generate one by concatenating
+     * the <code>matrix_id</code> and the <code>remote_id</code>. The delimiter
+     * used is a property on this store and can be modified.</p>
+     * <p>The structure of Entry objects means that it is efficient to select based
+     * off the 'id', 'matrix_id' or 'remote_id'. Additional indexes can be added
+     * manually.</p>
+     * @constructor
+     * @param {Datastore} db The connected NEDB database instance
+     * @param {Object} opts Options for this store.
+     * @property {string} delimiter The delimiter between matrix and
```
Comment (Half-Shot) [inline src/components/room-bridge-store.ts:90]:
```suggestion
     * @param db The connected NEDB database instance
```

File: src/components/room-bridge-store.ts [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+ *
+ */
+
+import Datastore from "nedb";
+import { BridgeStore } from "./bridge-store";
+import { MatrixRoom, MatrixRoomData } from "../models/rooms/matrix";
+import { RemoteRoom } from "../models/rooms/remote";
+
+export class RoomBridgeStore extends BridgeStore {
+    private readonly delimiter = "    ";
```
Comment (Half-Shot) [inline src/components/room-bridge-store.ts:65]:
```suggestion
    public delimiter = "    ";
```

File: src/components/room-bridge-store.ts [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+     * allows spaces, you will need to change this.
+     */
+    constructor(db: Datastore) {
+        super(db);
+    }
+
+    /**
+     * Insert an entry, clobbering based on the ID of the entry.
+     * @param {RoomBridgeStore~Entry} entry
+     * @return {Promise}
```
Comment (Half-Shot) [inline src/components/room-bridge-store.ts:101]:
```suggestion
     * @param entry
```

File: src/components/room-bridge-store.ts [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+    public upsertEntry(entry: RoomBridgeStoreEntry) {
+        return this.upsert({
+            id: entry.id
+        }, RoomBridgeStoreEntry.serializeEntry(entry) as Record<string, unknown>);
+    }
+
+    /**
+     * Get an existing entry based on the provided entry ID.
+     * @param {String} id The ID of the entry to retrieve.
+     * @return {?RoomBridgeStore~Entry} A promise which resolves to the entry or null.
```
Comment (Half-Shot) [inline src/components/room-bridge-store.ts:112]:
```suggestion
     * @param id The ID of the entry to retrieve.
     * @return A promise which resolves to the entry or null.
```

File: src/components/room-bridge-store.ts [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+            id: id
+        }, this.convertTo((doc: RoomStoreEntryDoc) =>
+            new RoomBridgeStoreEntry(doc)
+        ));
+    }
+
+    /**
+     * Get a list of entries based on the matrix_id of each entry.
+     * @param {string} matrixId
+     * @return {RoomBridgeStore~Entry[]}
```
Comment (Half-Shot) [inline src/components/room-bridge-store.ts:125]:
```suggestion
```

**Review by Half-Shot — COMMENTED** [review]
missed things

**Pull request merged**